date/date.1
'\" t
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\" Sccsid @(#)date.1	1.16 (gritter) 1/24/05
.TH DATE 1 "1/24/05" "Heirloom Toolchest" "User Commands"
.SH NAME
date \- print or set the date
.SH SYNOPSIS
\fBdate\fR [\fB\-u\fR] [\fB+\fIformat\fR]
.br
\fBdate\fR [\fB\-u\fR] [\fB\-b\fR]
\fIHHMM\fR | \fImmddHHMM\fR[[\fIcc\fR]\fIyy\fR]
.br
\fBdate\fR \fB\-a\fR [\fB\-\fR]\fIsss.fff\fR
.SH DESCRIPTION
.B
Date
displays the current date if invoked without arguments
or with a
.BI + format
argument.
.PP
If the first argument does not start with
.BR + ,
.B date
sets the system time to the given value,
which may consist of
.sp
.RS
.TS
lfB l.
mm	month of year (01\^\(en\^12)
dd	day of month (01\^\(en\^31)
HH	hour (00\^\(en\^23)
MM	minute (00\^\(en\^59)
cc	first two digits of year
yy	second two digits of year
.TE
.RE
.PP
The following options alter the behavior of
.BR date :
.TP
\fB\-a\fR [\fI\-\fR]\fIsss.fff\fR
Adjust the system clock by
.I sss
seconds and
.I fff
milliseconds,
speeding up or,
with
.IR \-,
slowing down.
.TP
.B \-u
Use
.SM UTC
values for printing and setting the date.
.PP
The following option is an extension:
.TP
.B \-b
Causes
.I date
not to alter the `wtmp' file for the system time change.
.PP
An argument starting with
.B +
specifies a format string to print the current date.
A percent character
.B %
introduces a field descriptor with one or two following characters
which is replaced by a part of the current date,
as shown in the table below.
Other characters are printed literally.
A newline character is printed at the end of the output.
.TP
Field descriptors:
.TS
l3 l.
a	T{
locale-specific abbreviated weekday name (System\ III)
T}
.\" the following line implies the minimum table width
A	locale-specific full weekday name (SVID3, POSIX.2)
b	T{
locale-specific abbreviated month name (SVID3, POSIX.2)
T}
B	T{
locale-specific full month name (SVID3, POSIX.2)
T}
c	T{
locale-specific date and time (SVID3, POSIX.2)
T}
C	T{
century \(en 19 or 20 (POSIX.2) (\fB/usr/5bin/s42/date\fR, \fB/usr/5bin/posix/date\fR, \fB/usr/5bin/posix2001/date\fR)
.br
default output format (\fB/usr/5bin/date\fR)
T}
d	T{
day of month \(en 01 to 31 (System\ III)
T}
D	T{
date as mm/dd/yy (System\ III)
T}
e	T{
day of month \(en 1 to 31 (SVID3, POSIX.2)
T}
F	T{
date as yyyy\-mm\-dd
T}
g	T{
week-based year within century \(en 00 to 99
T}
G	T{
week-based year including century \(en 0000 to 9999
T}
h	T{
same as %b (System\ III)
T}
H	T{
hour \(en 00 to 23 (System\ III)
T}
I	T{
hour \(en 01 to 12 (SVID3, POSIX.2)
T}
j	T{
day of year \(en 001 to 366 (System\ III)
T}
m	T{
month of year \(en 01 to 12 (System\ III)
T}
M	T{
minute \(en 00 to 59 (System\ III)
T}
n	T{
newline character (System\ III)
T}
N	T{
default output format
T}
p	T{
locale-specific AM\^/\^PM string (SVID3, POSIX.2)
T}
r	T{
locale-specific time using AM\^/\^PM notation (System\ III)
T}
R	T{
same as %H:%M (SVID3)
T}
S	T{
second \(en 00 to 61 (System\ III)
T}
t	T{
tab character (System\ III)
T}
T	T{
time as HH:MM:SS (System\ III)
T}
u	T{
weekday \(en 1 (Monday) to 7 (POSIX.2)
T}
U	T{
week of year (Sunday as first day of week) \(en 00 to 53 (SVID3, POSIX.2)
T}
V	T{
week of year (Monday as first day of week) \(en 01 to 53 (POSIX.2)
T}
w	T{
weekday \(en 0 (Sunday) to 6 (System\ III)
T}
W	T{
week of year (Monday as first day of week) \(en 00 to 53 (SVID3, POSIX.2)
T}
x	T{
locale-specific date (POSIX.2)
T}
X	T{
locale-specific time (SVID3, POSIX.2)
T}
y	T{
last two digits of year \(en 00 to 99 (System\ III)
T}
Y	T{
four-digit year \(en 1970 to 2037 (SVID3, POSIX.2)
T}
z	T{
numeric timezone
T}
Z	T{
timezone name (SVID3, POSIX.2)
T}
%	T{
percent sign (System\ III)
T}
Ec	T{
alternative date and time (POSIX.2)
T}
EC	T{
alternative base year (period) (POSIX.2)
T}
Ex	T{
alternative date (POSIX.2)
T}
EX	T{
alternative time (SUSv2)
T}
Ey	T{
offset from %EC in years (POSIX.2)
T}
EY	T{
full alternative year (POSIX.2)
T}
Od	T{
day of month filled with leading zero using alternative numerals (POSIX.2)
T}
Oe	T{
day of month filled with leading space using alternative numerals (POSIX.2)
T}
OH	T{
24-hour clok using alternative numerals (POSIX.2)
T}
OI	T{
12-hour clock using alternative numerals (POSIX.2)
T}
Om	T{
month of year using alternative numerals (POSIX.2)
T}
OM	T{
minute using alternative numerals (POSIX.2)
T}
OS	T{
second using alternative numerals (POSIX.2)
T}
Ou	T{
weekday (Monday = 1) using alternative numerals (POSIX.2)
T}
OU	T{
week of year (cf. %U) using alternative numerals (POSIX.2)
T}
OV	T{
week of year (cf. %V) using alternative numerals (POSIX.2)
T}
Ow	T{
weekday (Sunday = 0) using alternative numerals (POSIX.2)
T}
OW	T{
week of year (cf. %W) using alternative numerals (POSIX.2)
T}
Oy	T{
year (offset from %C) using alternative numerals (POSIX.2)
T}
.TE
.TP
Compatibility references:
.TS
l l.
.\" the following line implies the minimum table width
System III	AT&T System III manual page date(1).\ \ \ \ \ \ 
SVID3	T{
AT&T System V Interface Definition, Third Edition, Volume II, 1992,
date(BU_CMD). Includes all field descriptors present in System III.
T}
POSIX.2	T{
ISO/IEC 9945-2:1993(E), 4.15.4.1\^\(en\^4.15.4.2, pp. 241\^\(en\^242.
Includes all field descriptors present in System III.
T}
SUSv2	T{
The Open Group,
Single \s-1UNIX\s0
Specification, Version 2, \(co 1997, date(XCU).
Includes all field descriptors present in POSIX.2.
T}
.TE
.PP
When no arguments are given,
the format defaults to
.RS
.sp
%a %b %e %H:%M:%S %Z %Y
.sp
.RE
in the C or POSIX locale, resulting in an output like
.RS
.sp
Wed Feb 26 17:48:54 CET 2003
.sp
.RE
.SH "ENVIRONMENT VARIABLES"
.TP
.B "LANG, LC_ALL"
See
.IR locale (7).
.TP
.B LC_TIME
Alters the strings printed when format arguments are given.
.TP
.B TZ
Timezone; if unset, the system default is used.
.SH "SEE ALSO"
adjtime(2),
stime(2),
strftime(3),
locale(7)
.SH NOTES
After setting the date,
the
.IR hwclock (8)
utility is called to adjust the hardware clock on Linux systems.


CHANGES
Release ...
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the ssafe() function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complying about it and it
  is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107

date/date.c
/*
 * date - write the date and time
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, June 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 * Copyright (c) 2023 - 2024: Luiz Antônio (takusuman)
 *
 * SPDX-Licence-Identifier: Zlib
 */

#define USED
#if defined (SUS)
static const char sccsid[] USED = "@(#)date_sus.sl	1.27 (gritter) 12/16/07";
#else
static const char sccsid[] USED = "@(#)date.sl	1.27 (gritter) 12/16/07";
#endif

#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<time.h>
#include	<locale.h>
#include	<utmpx.h>
#include	<langinfo.h>

static unsigned	errcnt;			/* count of errors */
static int	bflag;			/* boot flag: don't write wtmp */
static int	uflag;			/* use UTC */
static char	*progname;		/* argv[0] to main() */
static time_t	now;			/* current time */

static void
usage(void)
{
	fprintf(stderr,
	"usage: %s [-u] [+format] [hhmm | mmddhhmm[[cc]yy]] [-a sss.fff]\n",
		progname);
	exit(1);
}

static void
badconv(void)
{
	fprintf(stderr, "%s: bad conversion\n", progname);
	exit(1);
}

static int
atot(char *s, int m)
{
	char *x;
	int i;

	i = (int)strtol(s, &x, 10);
	if (*x != '\0' || i > m || i < 0 || *s == '-' || *s == '+')
		return -1;
	return i;
}

static time_t
timeop(char *op)
{
	struct tm *tp;

	tp = uflag ? gmtime(&now) : localtime(&now);
	tp->tm_isdst = -1;
	tp->tm_sec = 0;
	switch(strlen(op)) {
	case 12:
		if ((tp->tm_year = atot(&op[8], 30000)) < 1970)
			goto badtime;
		tp->tm_year -= 1900;
		goto bypass;
	case 10:
		if ((tp->tm_year = atot(&op[8], 99)) < 0)
			goto badtime;
		if (tp->tm_year < 69)
			tp->tm_year += 100;
bypass:
		op[8] = '\0';
		/*FALLTHRU*/
	case 8:
		if ((tp->tm_min = atot(&op[6], 59)) < 0)
			goto badtime;
		op[6] = '\0';
		if ((tp->tm_hour = atot(&op[4], 23)) < 0)
			goto badtime;
		op[4] = '\0';
		if ((tp->tm_mday = atot(&op[2], 31)) <= 0)
			goto badtime;
		op[2] = '\0';
		if ((tp->tm_mon = atot(op, 12)) <= 0)
			goto badtime;
		tp->tm_mon--;
		break;
	case 4:
		if ((tp->tm_min = atot(&op[2], 59)) < 0)
			goto badtime;
		op[2] = '\0';
		if ((tp->tm_hour = atot(op, 23)) < 0)
			goto badtime;
		break;
	default:
		goto badtime;
	}
	return mktime(tp);
badtime:
	return (time_t)-1;
}

#if defined (__FreeBSD__) || defined (__NetBSD__) || defined (__OpenBSD__) || \
	defined (__DragonFly__) || defined (__APPLE__) || __GLIBC__ >= 2 && \
	__GLIBC_MINOR__ >= 31
static int
stime(time_t *t)
{
	struct timeval	tv;

	tv.tv_sec = *t;
	tv.tv_usec = 0;

	return settimeofday(&tv, NULL);
}
#endif	/* __FreeBSD__, __NetBSD__, __OpenBSD__, __DragonFly__, __APPLE__, GLIBC above 2.35*/

static void
settime(char *op)
{
	/*
	 * Since it's not guaranteed that (utmpx).ut_tv
	 * is a struct of the type timeval, we will be
	 * recording the time from gettimeofday(2) on
	 * a temporary struct of the type timeval and
	 * then copying the values to the ".ut_tv".
	 */
	struct utmpx before, after;
	struct timeval tv_tmp;
	const char wtmpxfile[] = "/var/log/wtmp";
	time_t newtime;

	memset(&before, 0, sizeof before);
	memset(&after, 0, sizeof after);
	before.ut_type = OLD_TIME;
	after.ut_type = NEW_TIME;
	strcpy(before.ut_line, "old time");
	strcpy(after.ut_line, "new time");
	if ((newtime = timeop(op)) == (time_t)-1)
		badconv();
	gettimeofday(&tv_tmp, NULL);
	before.ut_tv.tv_sec = tv_tmp.tv_sec;
	before.ut_tv.tv_usec = tv_tmp.tv_usec;
	/* Clean it for (struct utmpx)after. */
	memset(&tv_tmp, 0, sizeof tv_tmp);
	if (stime(&newtime) < 0) {
		fprintf(stderr, "%s: no permission\n", progname);
		exit(1);
	}
	gettimeofday(&tv_tmp, NULL);
	after.ut_tv.tv_sec = tv_tmp.tv_sec;
	after.ut_tv.tv_usec = tv_tmp.tv_usec;
#ifdef	__linux__
	system("/sbin/hwclock -w >/dev/null 2>&1");
#endif	/* __linux__ */
	if (bflag == 0) {
		pututxline(&before);
		pututxline(&after);
#if !defined (__hpux) && !defined (_AIX) && !defined(__APPLE__)
		updwtmpx(wtmpxfile, &before);
		updwtmpx(wtmpxfile, &after);
#endif	/* !__hpux, !__AIX, !__APPLE__ */
	}
	exit(0);
}

static void
printtime(const char *cp)
{
	struct tm *tp;
	char fmt[256];
	char *fp;
	char buf[256];
	int mod;
	char	*date_fmt;

	tp = uflag ? gmtime(&now) : localtime(&now);
	while (*cp != '\0') {
		if (*cp == '%') {
			const char	valid[] = "aAbBc"
#ifdef	SUS
					"C"
#endif	/* SUS */
					"dDeFgGhHIjmMnprRStTuUVwWxXyYZ%";
			fp = fmt;
			*fp++ = *cp++;
			if (*cp == 'E' || *cp == 'O')
				mod = *cp++;
			else
				mod = '\0';
			if (*cp == '\0') {
				putchar('%');
				goto end;
			}
			if (!strchr(valid, *cp)) {
				switch (*cp) {
				case 'N':
#ifndef	SUS
				case 'C':
#endif	/* !SUS */
#ifdef	_DATE_FMT
					if ((date_fmt = nl_langinfo(_DATE_FMT))
							== NULL ||
							*date_fmt == '\0')
#endif	/* _DATE_FMT */
						date_fmt =
						"%a %b %e %H:%M:%S %Z %Y";
					strcpy(fmt, date_fmt);
					break;
				case 'E':
				case 'O':
					goto next;
				case 'z': {
						  int hour, min;
#if defined (__FreeBSD__) || defined (__OpenBSD__) || defined (__DragonFly__) \
	|| defined (__APPLE__)
						  long	timezone;
						  timezone = -tp->tm_gmtoff;
#endif	/* __FreeBSD__, __OpenBSD__, __DragonFly__, __APPLE__ */

						  hour = -timezone / 3600;
						  min = (timezone / 60) % 60;
						  if (tp->tm_isdst > 0)
							  hour++;
						  fp = fmt;
						  if (hour <= 0) {
							  *fp++ = '-';
							  hour = -hour;
						  }
						  sprintf(fp, "%02d%02d",
								  hour, min);
					  }
					break;
				default:
					*fp++ = '%';
					if (mod)
						*fp++ = mod;
					*fp++ = *cp;
					*fp++ = '\0';
				}
			} else {
				if ((mod == 'E' && strchr("cCxXyY", *cp)) ||
						(mod == 'O' &&
						 strchr("deHImMSuUVwWy", *cp)))
					*fp++ = mod;
				*fp++ = *cp;
				*fp = '\0';
			}
			strftime(buf, sizeof buf, fmt, tp);
			fputs(buf, stdout);
		} else
			putchar(*cp);
next:
		cp++;
	}
end:
	putchar('\n');
}

static void
adjust(char *op)
{
	struct timeval tv;
	char *cp;

	if ((cp = strchr(op, '.')) != NULL) {
		if ((tv.tv_usec = atot(&cp[1], 999)) < 0)
			usage();
		tv.tv_usec *= 1000;
		*cp = '\0';
	}
	if ((tv.tv_sec = atot(*op == '-' ? &op[1] : op, 999)) < 0)
		usage();
	if (*op == '-') {
		tv.tv_sec = -tv.tv_sec;
		tv.tv_usec = -tv.tv_usec;
	}
	if (adjtime(&tv, 0) < 0) {
		fprintf(stderr, "%s: Failed to adjust date: %s\n", progname,
				strerror(errno));
		exit(1);
	}
	return;
}

int
main(int argc, char **argv)
{
	const char optstring[] = ":a:bu";
	char *adjustment = NULL;
	int i;

	time(&now);
	progname = basename(argv[0]);
	setlocale(LC_TIME, "");
#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'a':
			adjustment = optarg;
			break;
		case 'b':
			bflag = 1;
			break;
		case 'u':
			uflag = 1;
			putenv("TZ=GMT");
			tzset();
			break;
		case ':':
			usage();
			/*NOTREACHED*/
		default:
			/*fprintf(stderr, "%s: no TOY clock\n", progname);
			exit(2);*/
			badconv();
		}
	}
	if (adjustment) {
		if (uflag)
			badconv();
		adjust(adjustment);
	} else if (argv[optind] && argv[optind][0] == '+') {
		if (argv[optind][1] == '\0')
			goto dfl;
		printtime(&argv[optind][1]);
	} else if (argv[optind])
		settime(argv[optind]);
	else
	dfl:	printtime("%N");
	return errcnt;
}
